# Support Vector Machine (SVM)
## Conceito
O SVM é um modelo que tenta separar classes por um hiperplano, mas ele tenta otimizar por onde essa linha deve percorrer calculando a distância da reta aos pontos mais próximos dela, a margem do SVM é a distância entre a reta calculada com os dados mais próximos
<img src="Imagens/SVM/SVM.png"  width="800"/>

O SVM também funciona em planos 3D, ele calcula o melhor plano para classificar os dados, voltando e quando voltado ao domínio 2D permite uma maior complexidade em categorizar os dados

<img src="Imagens/SVM/SVM2.png"  width="800"/>

## Vantagens
    - Funciona melhor quando tem uma margem bem definida
    - Lida bem com espaços de altas dimensões
    - Possui eficiência na memória


## Desvantagens
    - Não é bom com grandes datasets
    - Requer um pré processamento dos dados para afinar os parâmetros


## Classes de Problemas com melhores resultados:

SVM permite complexas fronteiras na separação de dados mesmo com dados com baixa dimensão, funcionam bem tanto em baixa quanto alta dimensões mas não com um grande dataset.

## Exemplo

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

### Pimeiro Passo: Obter os dados

In [3]:
main = pd.read_csv('diabetes.csv')

In [4]:
main.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

### Segundo Passo: Divisão treino-teste

In [5]:
from sklearn.model_selection import train_test_split

In [6]:
x = main.drop('Outcome',axis=1)
y = main['Outcome']
x_train, x_test, y_train, y_test = train_test_split(x,y,test_size=0.3, random_state=101)

### Terceiro passo: Treinar o modelo

In [7]:
from sklearn.svm import SVC

In [8]:
svc_model= SVC()

In [9]:
svc_model.fit(x_train,y_train)

SVC()

Quarto Passo: Avaliar o modelo

In [11]:
pred = svc_model.predict(x_test)

In [12]:
from sklearn.metrics import classification_report,confusion_matrix

In [14]:
print(confusion_matrix(y_test,pred))

[[139  11]
 [ 40  41]]


In [15]:
print(classification_report(y_test,pred))

              precision    recall  f1-score   support

           0       0.78      0.93      0.84       150
           1       0.79      0.51      0.62        81

    accuracy                           0.78       231
   macro avg       0.78      0.72      0.73       231
weighted avg       0.78      0.78      0.76       231



SVM deu um modelo de 78% de Precisão, bem parecido com o KNN